### ISSUE TITLE
Push for release 0.7.0

### ISSUE THREAD

**Issue by Thrameos** [jpype-project/jpype#419]
Update on #371.

Okay I think it is past time to cut a 0.7.0.   I have been running it only my main project for about a month.  I think it is getting close to "production" quality.  For my purposes a release candidate should:

1. Have no major issues that result in a failure.
2. Introduces no regrettable API choices would force us to remove compatibility at future points.
3. Only breaks previous documented API for issues of gross problems.  That is API decisions that locked us into an implementation that leads to problems.

Based on my testing thus far, we are pretty close to that state.

The only crash I have experienced was a race condition between the python interrupt handler and the C++ layer.  It is a known issue that if certain JNI operations happen while python is proceeding, we get into an unknown state.  Everything is pretty well armed right now, but we are depending on the global lock to make that happen.  Every crash I have had involved issuing a control C to break a java operation which returned us to python and then tried to access java in that state.  Obvious this one is pretty hard to replicate, and not worth holding a release up.

As to introducing regrettable behavior,  I have been pretty consistent about removing behaviors rather than adding them.  The few that have been introduced are more type conversions and flexibility with existing elements.   The main API that was introduced was the proxy and customizer annotations, neither of which I see as regrettable currently.  On type conversions, I have one odd behavior in which a JObject double wrap on an array placed into a generic list object gave me back an unexpected result.   Our testing code is getting more complete, but there are still enough holes that someone could depend on something converting one way that may change without tripping the testing frame work.  I feel at this time the memory compiler would fall into the regrettable behavior.  Though I don't feel there is anything wrong with the code, it has marginal utility as it only works with the JDK version of the JVM.  

As to regrettable stuff that was removed we have a boat load of them:
1) JavaClass and JavaObject base classes.   These were introduced to make it possible to test instances.  They appears in the change logs, but have not otherwise been documented.  They have been incorporated into the existing factory classes and thus serve no purpose.  As undocumented extra symbols they have been removed.
2) JException and the exception tree.  Exception handling in 0.6.3 requires catching a weird second tree of python objects rather than the actually exception.  This disunity was problematic at the C++ layer and unnecessary.   I have tried to keep the exception handling similar, but removing a whole second tree.  We could make this somewhat more compatible by pretending to access the previously buried copy of the java exception.
3) The wrapper classes as distinct from class instances.  Unifying all types as `__java_value__` objects using duck typing made a huge difference in the complexity of the core.   Any code that was using stuff under the hood on those types will break.
4) Customizers is another area of major change.  The old process was not documented so I have little regrets about killing it.  It also did not scale well as every customizer needed to probe every class wrapper created.
5) the property customizer was the worst of them.  This global customizer was introduced late in the loading process after other classes were loaded so it only applied user loaded classes.  It adds a second way to call get/set which really violates Python philosophy there should be one clear way to perform an action.  We have relegated it to an optional module jpype.beans.
6) Access to private members.   This unexpected behavior was a bug in the C++ core, but someone may be using it. 
7) The reflect package.  This existed before we had access to the full `class_` object.  It has been deprecated rather than removed as it was not doing harm as an optional module.

 If anyone is interested in arguing that something they use is missing from 0.7.0 candidate, now is the time.  Otherwise, we are nearly ready for a release.

**Comment by Thrameos** [comment]
@marscher Okay the future has arrived.  Everything that was needed to make a 0.7 release has been completed.  Thus the last remaining steps need to happen.  

[ ] Build a test release of the dist
[ ] Uninstall the JPype from the machine
[ ] Try to build from the test dist
[ ] Lather, Rinse, Repeat until the dist actually works
[ ] Put a pull request which fixes this issue and all the dist related problems.

Unfortunately, this is the one part that I have never done with a Python project.  Thus I will have to look over the setup.py instructions until I figure out how to make it happen.  Once that is complete we should be ready to release this.

**Comment by marscher** [comment]
python setup.py sdist will create a source distribution. If you need to ship additional files, take a look at the file MANIFEST.in

**Comment by marscher** [comment]
Actually we should have a job on Travis performing a sdist and a pip install of the latter, prior running the tests. I can add that, so you can just take it from this source then.

**Comment by marscher** [comment]
It already performs the described protocol.

**Comment by marscher** [comment]
There are a lot of merge conflicts with master. How should we proceed with them?

**Comment by Thrameos** [comment]
I was hoping to rename master to 0.6.3 and rename devel as master, though it may be very naive of me.  Due to problems in the way that the devel came about there are a lot of things that don't sync as far as histories.   If you want me to try to make a PR from devel to master (resolving those internal problems) I can attempt it.

**Comment by marscher** [comment]
The branch renaming solution has an appealing simplicity :)

**Comment by Thrameos** [comment]
I will leave this decision to you.  Branch renaming will break all those copies of master that are out there.  Creating the PR with everything resolved will take about 2 hours which I should be able to do tonight.  Just point me at which path we want and I will make it happen.

**Comment by marscher** [comment]
You're right about the branch renaming, however master an devel share a 
common ancestor. So I'm not sure of the actual breakage.
If you think it would be more clean to do a merge, I would prefer it.

**Comment by marscher** [comment]
documentation is in place: https://jpype.readthedocs.io/en/devel/

Installing via pip also works:
```
$ pip install git+https://github.com/jpype-project/jpype@devel#egg=jpype1
```

Is it now mandatory to have a jdk and ant installed to install jpype? Could we circumvent this by shipping a jar containing byte codes?

**Comment by marscher** [comment]
Do you mind removing the section "tested on" in the install instructions? It is pretty much outdated. Stating which Java versions are supported is more of interest (if it is not yet present).

**Comment by Thrameos** [comment]
Yes, we can have the packager put the current jar into the binary distribution.  But I will need to work on the pattern so it properly builds if the jar is out-of-date.  Having to specify ant everytime gets pretty old.

I will touch up the tested on section when I get back to this later.

**Comment by Thrameos** [comment]
Okay there is a lot more editing needed in the install.  Some of these limitations were solved before I got here "unable to access a field or method if it conflicts with a python keyword."

**Comment by Thrameos** [comment]
@marscher I think we are likely ready for release 0.7.0

**Comment by marscher** [comment]
Very well, I've tested 0.7 locally and asked two maintainers of jpype dependent projects to do the same for their software. I think it would be polite to give them a couple of days to respond.
We can then upload the release to pypi. Or do you prefer doing some cycles of release candidates?

**Comment by Thrameos** [comment]
I am fine with whatever road you chose.  I debated pulling down a few of the largest jpype codes and seeing how the ran, but github wouldn't sort for me and several I did find were asian language.  I feel that we have a release candidate today.  Assuming no one objects it should be safe to push out this Friday.

I am available for the next few months to help with answering the mail and fixing issues.  In the meantime I am working on the project list for 0.8 series.
